coreneuron/io/phase1.cpp

```diff
@@ -0,0 +1,170 @@
+#include <cassert>
+
+#include "coreneuron/nrniv/nrniv_decl.h"
+#include "coreneuron/sim/multicore.hpp"
+#include "coreneuron/io/phase1.hpp"
+#include "coreneuron/utils/nrnoc_aux.hpp"
+
+int (*nrn2core_get_dat1_)(int tid,
+                          int& n_presyn,
+                          int& n_netcon,
+                          int*& output_gid,
+                          int*& netcon_srcgid);
+
+namespace coreneuron {
+// no gid in any file can be greater than maxgid. maxgid will be set so that
+// maxgid * nrn_setup_multiple < 0x7fffffff
+static int maxgid = 0x7fffffff / nrn_setup_multiple;
+
+void Phase1::read_direct(int thread_id) {
+    int* output_gids;
+    int* netcon_srcgid;
+    int n_presyn;
+    int n_netcon;
+    int valid =
+        (*nrn2core_get_dat1_)(thread_id, n_presyn, n_netcon, output_gids, netcon_srcgid);
+    if (!valid) {
+        return;
+    }
+
+    this->output_gids = std::vector<int>(output_gids, output_gids + n_presyn);
+    delete[] output_gids;
+    this->netcon_srcgids = std::vector<int>(netcon_srcgid, netcon_srcgid + n_netcon);
+    delete[] netcon_srcgid;
+}
+
+void Phase1::read_file(FileHandler& F) {
+    assert(!F.fail());
+    int n_presyn = F.read_int();  /// Number of PreSyn-s in NrnThread nt
+    int n_netcon = F.read_int();  /// Number of NetCon-s in NrnThread nt
+
+    this->output_gids = F.read_vector<int>(n_presyn);
+    this->netcon_srcgids = F.read_vector<int>(n_netcon);
+
+    F.close();
+}
+
+void Phase1::shift_gids(int imult) {
+    int zz = imult * maxgid;  // offset for each gid
+
+    // offset the (non-negative) gids according to multiple
+    // make sure everything fits into gid space.
+    for (auto& gid: this->output_gids) {
+        if (gid >= 0) {
+            nrn_assert(gid < maxgid);
+            gid += zz;
+        }
+    }
+
+    for (auto& srcgid: this->netcon_srcgids) {
+        if (srcgid >= 0) {
+            nrn_assert(srcgid < maxgid);
+            srcgid += zz;
+        }
+    }
+}
+
+void Phase1::add_extracon(NrnThread& nt, int imult) {
```

Comment:
note for ourselves : to be removed in https://github.com/BlueBrain/CoreNeuron/issues/291